Source/WebCore/css/calc/CSSCalcOperationNode.cpp

```diff
@@ -536,46 +553,38 @@ RefPtr<CSSCalcOperationNode> CSSCalcOperationNode::createStep(CalcOperator op, V
     if (values.size() != 2)
         return nullptr;
 
-    if (values[0]->category() != values[1]->category()) {
-        LOG_WITH_STREAM(Calc, stream << "Failed to create stepped value node because unable to determine category from " << prettyPrintNodes(values));
-        return nullptr;
-    }
-    return adoptRef(new CSSCalcOperationNode(values[0]->category(), op, WTFMove(values)));
-}
+    if (auto category = resolvedTypeForStep(values[0]->category(), values[1]->category()))
+        return adoptRef(new CSSCalcOperationNode(*category, op, WTFMove(values)));
 
-static bool validateRoundChildren(Vector<Ref<CSSCalcExpressionNode>>& values)
-{
-    // for 3 children 1st node must be round constant
-    if (values.size() == 3) {
-        if (!is<CSSCalcOperationNode>(values[0]) || !(downcast<CSSCalcOperationNode>(values[0].get()).isRoundOperation()))
-            return false;
-    }
-    // for 2 children should not have round constant anywhere but first node of 3
-    for (size_t i = values.size() == 2 ? 0 : 1; i < values.size(); i++) {
-        if (is<CSSCalcOperationNode>(values[i])) {
-            if (downcast<CSSCalcOperationNode>(values[i].get()).isRoundConstant())
-                return false;
-        }
-    }
-    // check that two categories of numerical values are the same
-    return values.rbegin()[1]->category() == values.rbegin()[0]->category();
-    
+    LOG_WITH_STREAM(Calc, stream << "Failed to create stepped value node because unable to determine category from " << prettyPrintNodes(values));
+    return nullptr;
 }
 
 RefPtr<CSSCalcOperationNode> CSSCalcOperationNode::createRound(Vector<Ref<CSSCalcExpressionNode>>&& values)
 {
     if (values.size() != 2 && values.size() != 3)
         return nullptr;
-    
-    if (!validateRoundChildren(values)) {
-        LOG_WITH_STREAM(Calc, stream << "Failed to create round node because unable to determine category from " << prettyPrintNodes(values));
-        return nullptr;
+
+    auto roundOperationFromNode = [](Ref<CSSCalcExpressionNode>& node) -> std::optional<CalcOperator> {
+        if (auto value = dynamicDowncast<CSSCalcOperationNode>(node.get()); value && value->isRoundConstant())
+            return value->calcOperator();
+        return { };
+    };
+
+    auto roundOperation = CalcOperator::Nearest;
+
+    if (values.size() == 3) {
+        if (auto operation = roundOperationFromNode(values[0])) {
+            roundOperation = *operation;
+            values.remove(0);
+        } else
+            return nullptr;
     }
-    
-    CalcOperator roundType = values.size() == 2 ?  CalcOperator::Nearest : downcast<CSSCalcOperationNode>(values[0].get()).calcOperator();
-    if (values.size() == 3)
-        values.remove(0);
-    return adoptRef(new CSSCalcOperationNode(values.rbegin()[0]->category(), roundType, WTFMove(values)));
+
+    if (roundOperationFromNode(values[0]) || roundOperationFromNode(values[1]))
+        return nullptr;
+
```

Comment:
Can we write a test for this? and figure out if this bit is necessary?

Reply:
I think it is not necessary in the end but I will confirm.

Reply:
Turns out it is needed, and without it css/css-values/round-mod-rem-invalid.html fails.